aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,606 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_types::vm_status::StatusCode;
+use better_any::{Tid, TidAble};
+use move_deps::{
+    move_binary_format::errors::{PartialVMError, PartialVMResult},
+    move_core_types::{account_address::AccountAddress, gas_schedule::GasCost},
+    move_table_extension::{TableHandle, TableResolver},
+    move_vm_runtime::{
+        native_functions,
+        native_functions::{NativeContext, NativeFunctionTable},
+    },
+    move_vm_types::{
+        loaded_data::runtime_types::Type,
+        natives::function::NativeResult,
+        pop_arg,
+        values::{IntegerValue, Reference, Struct, StructRef, Value},
+    },
+};
+use smallvec::smallvec;
+use std::{
+    cell::RefCell,
+    collections::{btree_map::Entry, BTreeMap, BTreeSet, VecDeque},
+};
+use tiny_keccak::{Hasher, Sha3};
+
+use crate::delta_ext::{addition, deserialize, serialize, subtraction, DeltaOp};
+
+/// State of an aggregator: `Data` means that aggregator stores an exact value,
+/// `PositiveDelta` means that actual value is not known but must be added later
+#[derive(Clone, Copy, PartialEq)]
+pub enum AggregatorState {
+    Data,
+    PositiveDelta,
+}
+
+/// Internal aggregator data structure.
+struct Aggregator {
+    // Uniquely identifies aggregator in `AggregatorTable`.
+    key: u128,
+    // Describes state of aggregator: data or delta.
+    state: AggregatorState,
+    // Describes the value of data or delta.
+    value: u128,
+    // Postondition value, exceeding it aggregator overflows.
+    limit: u128,
+}
+
+impl Aggregator {
+    /// Implements logic for adding to an aggregator.
+    fn add(&mut self, value: u128) -> PartialVMResult<()> {
+        // At this point, aggregator holds a positive delta or knows the value.
+        // Hence, we can add, of course checking overflow condition.
+        let result = addition(self.value, value, self.limit)?;
+        self.value = result;
+        Ok(())
+    }
+
+    /// Implements logic for subtracting from an aggregator.
+    fn sub(&mut self, value: u128) -> PartialVMResult<()> {
+        match self.state {
+            AggregatorState::Data => {
+                // Aggregator knows the value, therefore we can subtract
+                // checking we don't drop below zero.
+                let result = subtraction(self.value, value)?;
+                self.value = result;
+                Ok(())
+            }
+            // Since `sub` is a barrier, we never encounter delta during
+            // subtraction. This will change in future.
+            // TODO: implement this properly once `sub` stops being a barrier.
+            AggregatorState::PositiveDelta => {
+                unreachable!("subtraction always materializes the value")
+            }
+        }
+    }
+
+    /// Materializes the value of the aggregator, changing its state.
+    fn materialize(
```

Comment:
What happens if this function fails? Would this fail all transactions that use an aggregator?

Reply:
On error VM aborts the transaction, no aggregator changes (deltas) will be produced or applied.

Reply:
So if we use this for something in the critical flow such as transaction fees, this can run the risk of breaking all transactions? Is there any kind of graceful fallback we can do here to potentially prevent that from happening?

Reply:
Why would it break all transactions?

Say we are executing a transaction that was doing +1000 and it failed for some reason (overflow, etc.). Then the transaction is aborted, but the next one is ok to proceed.

`materialize` fails in 2 ways: 1) overflow - in this case a descriptive message is passed to VM and then to user 2) some storage error, e.g. applying delta to non-existing value. This latter case should not ever happen really, since when we create an aggregator we always produce a write of 0, so all subsequent access to storage should succeed.